### PR TITLE
1773 refactor rewards spec

### LINKED ISSUE #1773: 🚀 [Refactor] Refactor rewards spec to use apollo-server-testing
> [aonomike](https://github.com/aonomike) **Authored by [aonomike](https://github.com/aonomike)**
_<time datetime="2019-09-30T09:00:35Z" title="Monday, September 30th 2019, 11:00:35 am +02:00">Sep 30, 2019</time>_
_Closed <time datetime="2019-12-29T09:49:53Z" title="Sunday, December 29th 2019, 10:49:53 am +01:00">Dec 29, 2019</time>_
---

## :rocket: Refactor

All tests in the reward spec in file `src/schema/resolvers/rewards.spec.js` ar e failing because its not using `appollo-server-testing` as most of the tests have been refactored to. This should be refactored to use Apollo server testing

### Additional context

Make all the tests in `src/schema/resolvers/rewards.spec.js` pass

### PR DESCRIPTION
**Pull request by Tirokk**
> [aonomike](https://github.com/aonomike) **Authored by [aonomike](https://github.com/aonomike)**
_<time datetime="2019-10-12T14:41:28Z" title="Saturday, October 12th 2019, 4:41:28 pm +02:00">Oct 12, 2019</time>_
_Merged <time datetime="2019-10-14T14:32:24Z" title="Monday, October 14th 2019, 4:32:24 pm +02:00">Oct 14, 2019</time>_
---

## 🍰 Pullrequest

Refactor rewards spec to use apollo-server-testing
### Issues

- None

### Todo

- [X] None

### CODE REVIEW

File: backend/src/schema/resolvers/rewards.spec.js [hunk 1]
```diff
@@ -54,51 +72,61 @@ describe('rewards', () => {
...
       })
     })
 
     describe('authenticated admin', () => {
-      let client
       beforeEach(async () => {
-        const headers = await login({ email: 'admin@example.org', password: '1234' })
-        client = new GraphQLClient(host, { headers })
+        authenticatedUser = await administrator.toJson()
       })
```
Comment (Tirokk) [inline backend/src/schema/resolvers/rewards.spec.js:88]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-14T12:14:22Z" title="Monday, October 14th 2019, 2:14:22 pm +02:00">Oct 14, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -54,51 +72,56 @@ describe('rewards', () => {
     `
 
     describe('unauthenticated', () => {
-      let client
-
       it('throws authorization error', async () => {
-        client = new GraphQLClient(host)
-        await expect(client.request(mutation, variables)).rejects.toThrow('Not Authorised')
+        authenticatedUser = null
+        await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+          data: { reward: null },
+          errors: [{ message: 'Not Authorised!' }],
+        })
       })
     })
 
     describe('authenticated admin', () => {
-      let client
       beforeEach(async () => {
-        const headers = await login({ email: 'admin@example.org', password: '1234' })
-        client = new GraphQLClient(host, { headers })
+        authenticatedUser = await administrator.toJson()
       })
 
       describe('badge for id does not exist', () => {
         it('rejects with a telling error message', async () => {
           await expect(
-            client.request(mutation, {
-              ...variables,
-              from: 'bullshit',
-            }),
-          ).rejects.toThrow("Couldn't find a badge with that id")
+            mutate({ mutation, variables: { to: 'regular-user-id', from: 'ndiegogo_en_rhino' } }),
```


```suggestion
            mutate({ mutation, variables: { to: 'regular-user-id', from: 'non-existent-badge-id' } }),
```

File: backend/src/schema/resolvers/rewards.spec.js [hunk 2]
```diff
@@ -54,51 +72,61 @@ describe('rewards', () => {
...
-              from: 'bullshit',
+            mutate({
+              mutation: rewardMutation,
+              variables: { to: 'regular-user-id', from: 'non-existent-badge-id' },
             }),
-          ).rejects.toThrow("Couldn't find a badge with that id")
+          ).resolves.toMatchObject({
+            data: { reward: null },
+            errors: [{ message: "Couldn't find a badge with that id" }],
+          })
```
Comment (Tirokk) [inline backend/src/schema/resolvers/rewards.spec.js:99]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-14T12:16:14Z" title="Monday, October 14th 2019, 2:16:14 pm +02:00">Oct 14, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -54,51 +72,56 @@ describe('rewards', () => {
     `
 
     describe('unauthenticated', () => {
-      let client
-
       it('throws authorization error', async () => {
-        client = new GraphQLClient(host)
-        await expect(client.request(mutation, variables)).rejects.toThrow('Not Authorised')
+        authenticatedUser = null
+        await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+          data: { reward: null },
+          errors: [{ message: 'Not Authorised!' }],
+        })
       })
     })
 
     describe('authenticated admin', () => {
-      let client
       beforeEach(async () => {
-        const headers = await login({ email: 'admin@example.org', password: '1234' })
-        client = new GraphQLClient(host, { headers })
+        authenticatedUser = await administrator.toJson()
       })
 
       describe('badge for id does not exist', () => {
         it('rejects with a telling error message', async () => {
           await expect(
-            client.request(mutation, {
-              ...variables,
-              from: 'bullshit',
-            }),
-          ).rejects.toThrow("Couldn't find a badge with that id")
+            mutate({ mutation, variables: { to: 'regular-user-id', from: 'ndiegogo_en_rhino' } }),
+          ).resolves.toMatchObject({
+            data: { reward: null },
+            errors: [{ message: "Couldn't find a badge with that id" }],
+          })
         })
       })
 
       describe('user for id does not exist', () => {
         it('rejects with a telling error message', async () => {
           await expect(
-            client.request(mutation, {
-              ...variables,
-              to: 'bullshit',
+            mutate({
+              mutation,
+              variables: { to: 'non-existent-user-id', from: 'non-existent-badge' },
```


```suggestion
              variables: { to: 'non-existent-user-id', from: 'indiegogo_en_rhino' },
```

File: backend/src/schema/resolvers/rewards.spec.js [hunk 3]
```diff
@@ -108,77 +136,116 @@ describe('rewards', () => {
...
     describe('authenticated moderator', () => {
-      let client
       beforeEach(async () => {
-        const headers = await login({ email: 'moderator@example.org', password: '1234' })
-        client = new GraphQLClient(host, { headers })
+        authenticatedUser = moderator.toJson()
       })
 
-      describe('rewards bage to user', () => {
+      describe('rewards badge to user', () => {
```
Comment (Tirokk) [inline backend/src/schema/resolvers/rewards.spec.js:228]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-14T12:18:47Z" title="Monday, October 14th 2019, 2:18:47 pm +02:00">Oct 14, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -108,75 +131,114 @@ describe('rewards', () => {
         })
         const badges = [{ id: 'indiegogo_en_racoon' }, { id: 'indiegogo_en_rhino' }]
         const expected = {
-          reward: {
-            id: 'u1',
-            badges: expect.arrayContaining(badges),
+          data: {
+            reward: {
+              id: 'regular-user-id',
+              badges: expect.arrayContaining(badges),
+            },
           },
+          errors: undefined,
         }
-        await client.request(mutation, variables)
+        await mutate({
+          mutation,
+          variables: {
+            to: 'regular-user-id',
+            from: 'indiegogo_en_rhino',
+          },
+        })
         await expect(
-          client.request(mutation, {
-            ...variables,
-            from: 'indiegogo_en_racoon',
+          mutate({
+            mutation,
+            variables: {
+              to: 'regular-user-id',
+              from: 'indiegogo_en_racoon',
+            },
           }),
-        ).resolves.toEqual(expected)
+        ).resolves.toMatchObject(expected)
       })
 
       it('rewards the same badge as well to another user', async () => {
         const expected = {
-          reward: {
-            id: 'u2',
-            badges: [{ id: 'indiegogo_en_rhino' }],
+          data: {
+            reward: {
+              id: 'regular-user-2-id',
+              badges: [{ id: 'indiegogo_en_rhino' }],
+            },
           },
+          errors: undefined,
         }
+        await factory.create('User', {
+          id: 'regular-user-2-id',
+          email: 'regular2@email.com',
+        })
+        await mutate({
+          mutation,
+          variables,
+        })
         await expect(
-          client.request(mutation, {
-            ...variables,
-            to: 'u2',
+          mutate({
+            mutation,
+            variables: {
+              to: 'regular-user-2-id',
+              from: 'indiegogo_en_rhino',
+            },
           }),
-        ).resolves.toEqual(expected)
+        ).resolves.toMatchObject(expected)
       })
 
       it('creates no duplicate reward relationships', async () => {
-        await client.request(mutation, variables)
-        await client.request(mutation, variables)
+        await mutate({
+          mutation,
+          variables,
+        })
+        await mutate({
+          mutation,
+          variables,
+        })
 
-        const query = gql`
+        const userQuery = gql`
           {
-            User(id: "u1") {
+            User(id: "regular-user-id") {
               badgesCount
               badges {
                 id
               }
             }
           }
         `
-        const expected = { User: [{ badgesCount: 1, badges: [{ id: 'indiegogo_en_rhino' }] }] }
+        const expected = {
+          data: { User: [{ badgesCount: 1, badges: [{ id: 'indiegogo_en_rhino' }] }] },
+          errors: undefined,
+        }
 
-        await expect(client.request(query)).resolves.toEqual(expected)
+        await expect(query({ query: userQuery })).resolves.toMatchObject(expected)
       })
     })
 
     describe('authenticated moderator', () => {
-      let client
       beforeEach(async () => {
-        const headers = await login({ email: 'moderator@example.org', password: '1234' })
-        client = new GraphQLClient(host, { headers })
+        authenticatedUser = moderator.toJson()
       })
 
-      describe('rewards bage to user', () => {
+      describe('rewards badge to user', () => {
```


good catch :)

File: backend/src/schema/resolvers/rewards.spec.js [hunk 4]
```diff
@@ -191,58 +258,65 @@ describe('rewards', () => {
...
-            User(id: "u1") {
+            User(id: "regular-user-id") {
               badgesCount
               badges {
                 id
               }
             }
           }
         `
-        const expected = { User: [{ badgesCount: 1, badges: [{ id: 'indiegogo_en_rhino' }] }] }
```
Comment (Tirokk) [inline backend/src/schema/resolvers/rewards.spec.js:204]: > [mattwr18](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-14T12:22:27Z" title="Monday, October 14th 2019, 2:22:27 pm +02:00">Oct 14, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -191,58 +253,61 @@ describe('rewards', () => {
 
     describe('check test setup', () => {
       it('user has one badge', async () => {
-        const query = gql`
+        authenticatedUser = regularUser.toJson()
+        const userQuery = gql`
           {
-            User(id: "u1") {
+            User(id: "regular-user-id") {
               badgesCount
               badges {
                 id
               }
             }
           }
         `
-        const expected = { User: [{ badgesCount: 1, badges: [{ id: 'indiegogo_en_rhino' }] }] }
-        const client = new GraphQLClient(host)
-        await expect(client.request(query)).resolves.toEqual(expected)
+        const expected = {
+          data: { User: [{ badgesCount: 1, badges: [{ id: 'indiegogo_en_rhino' }] }] },
+          errors: undefined,
+        }
+        await expect(query({ query: userQuery })).resolves.toMatchObject(expected)
       })
     })
 
     describe('unauthenticated', () => {
-      let client
-
       it('throws authorization error', async () => {
-        client = new GraphQLClient(host)
-        await expect(client.request(mutation, variables)).rejects.toThrow('Not Authorised')
+        authenticatedUser = null
+        await expect(mutate({ mutation, variables })).resolves.toMatchObject({
```


don't be afraid to follow your instincts and be prescriptive with your variable naming...

**Review by Mogge — APPROVED** [review]
great job @aonomike!!

**Pull request closed without merging**